applications/serial_lte_modem/src/mqtt_c/slm_at_mqtt.c

```diff
@@ -217,41 +220,69 @@ static void mqtt_thread_fn(void *arg1, void *arg2, void *arg3)
 	while (true) {
 		if (!ctx.connected) {
 			LOG_WRN("MQTT disconnected");
+			err = 0;
 			break;
 		}
 		err = poll(&fds, 1, mqtt_keepalive_time_left(&client));
 		if (err < 0) {
 			LOG_ERR("ERROR: poll %d", errno);
+			err = -EIO;
 			break;
 		}
-		/* timeout or revent, send KEEPALIVE */
-		(void)mqtt_live(&client);
-
-		if ((fds.revents & POLLIN) == POLLIN) {
-			err = mqtt_input(&client);
-			if (err != 0) {
-				LOG_ERR("ERROR: mqtt_input %d", err);
-				mqtt_abort(&client);
-				break;
-			}
-		}
+		/* MQTT v3.1.1: Note that a Server is permitted to disconnect a Client that it
+		 * determines to be inactive or non-responsive at any time, regardless of the
+		 * Keep Alive value provided by that Client.
+		 */
 		if ((fds.revents & POLLERR) == POLLERR) {
 			LOG_ERR("POLLERR");
-			mqtt_abort(&client);
+			err = -ENETDOWN;
 			break;
 		}
 		if ((fds.revents & POLLHUP) == POLLHUP) {
 			LOG_ERR("POLLHUP");
-			mqtt_abort(&client);
+			err = -ECONNRESET;
```

Comment:
POLLHUP indicates gracefull termination, err should be 0.

Reply:
cannot be 0 as need to do abort if unexpected down below.

Reply:
I understand where you are coming from, but the error is misleading in this case. I think that even -ENOTCONN would be better. Best would be to return 0.

Reply:
Original MQTT poll behavior was copied from TCP/UDP client

Reply:
Yes, the same behavior is in TCP server, and it is wrong there as well.

Reply:
I don't think there is a need to discuss further what error code to return in this PR as not serving the main purpose.
But it can be done in a new task, to review all the poll threads and other scenarios on the return errno in SLM-wide

Reply:
Can be done as a separate change.